File: app/utils/kelly.py
"""Shared utilities for Kelly Criterion calculations."""

from dataclasses import dataclass
from typing import Iterable, Any

import numpy as np


@dataclass(frozen=True)
class KellyMetrics:
    """Container for Kelly Criterion related outputs."""

    fraction: float
    percent: float
    win_rate: float
    payoff_ratio: float
    avg_win: float
    avg_loss: float


_ZERO_METRICS = KellyMetrics(
    fraction=0.0,
    percent=0.0,
    win_rate=0.0,
    payoff_ratio=0.0,
    avg_win=0.0,
    avg_loss=0.0,
)


def _extract_pl(trade: Any, pl_key: str) -> float:
    """Best-effort extraction of profit/loss from trade-like objects."""
    if isinstance(trade, dict):
        value = trade.get(pl_key, 0)  # type: ignore[index]
    else:
        value = getattr(trade, pl_key, 0)

    try:
        return float(value)
    except (TypeError, ValueError):
        return 0.0


def calculate_kelly_metrics(trades: Iterable[Any], pl_key: str = "pl") -> KellyMetrics:
    """Return Kelly Criterion metrics for an iterable of trades.

    Trades may be dictionaries or objects with a `pl` attribute. When insufficient
    information is available (no wins, no losses, or zero denominator), the
    function returns zeroed metrics.
    """
    trades_list = list(trades)

    if not trades_list:
        return _ZERO_METRICS

    wins = []
    losses = []

    for trade in trades_list:
        pl = _extract_pl(trade, pl_key)
        if pl > 0:
            wins.append(pl)
        elif pl < 0:
            losses.append(abs(pl))

    if not wins or not losses:
        return _ZERO_METRICS

    total_trades = len(trades_list)
    win_rate = len(wins) / total_trades
    avg_win = float(np.mean(wins))
    avg_loss = float(np.mean(losses))

    if avg_loss == 0:
        return _ZERO_METRICS

    payoff_ratio = avg_win / avg_loss
    loss_rate = 1 - win_rate
    kelly_fraction = (payoff_ratio * win_rate - loss_rate) / payoff_ratio
    kelly_percent = kelly_fraction * 100

    return KellyMetrics(
        fraction=kelly_fraction,
        percent=kelly_percent,
        win_rate=win_rate,
        payoff_ratio=payoff_ratio,
        avg_win=avg_win,
        avg_loss=avg_loss,
    )

File: lib/calculations/kelly.ts
/**
 * Kelly Criterion calculations for position sizing
 */

import { Trade } from "@/lib/models/trade";

export interface KellyMetrics {
  fraction: number;
  percent: number;
  winRate: number;
  payoffRatio: number;
  avgWin: number;
  avgLoss: number;
  hasValidKelly: boolean; // Indicates if Kelly can be calculated
}

const ZERO_METRICS: KellyMetrics = {
  fraction: 0,
  percent: 0,
  winRate: 0,
  payoffRatio: 0,
  avgWin: 0,
  avgLoss: 0,
  hasValidKelly: false,
};

/**
 * Calculate Kelly Criterion metrics for a set of trades
 *
 * Returns metrics with actual win rate but zero Kelly fraction if insufficient data
 * (no wins, no losses, or zero denominator)
 */
export function calculateKellyMetrics(trades: Trade[]): KellyMetrics {
  if (trades.length === 0) {
    return ZERO_METRICS;
  }

  const wins: number[] = [];
  const losses: number[] = [];

  for (const trade of trades) {
    const pl = trade.pl || 0;
    if (pl > 0) {
      wins.push(pl);
    } else if (pl < 0) {
      losses.push(Math.abs(pl));
    }
  }

  const totalTrades = trades.length;
  const winRate = wins.length / totalTrades;
  const avgWin = wins.length > 0
    ? wins.reduce((sum, val) => sum + val, 0) / wins.length
    : 0;
  const avgLoss = losses.length > 0
    ? losses.reduce((sum, val) => sum + val, 0) / losses.length
    : 0;

  // Check if we can calculate valid Kelly metrics
  const hasValidKelly = wins.length > 0 && losses.length > 0 && avgLoss > 0;

  if (!hasValidKelly) {
    // Return actual stats but with zero Kelly fraction
    return {
      fraction: 0,
      percent: 0,
      winRate,
      payoffRatio: avgLoss > 0 ? avgWin / avgLoss : 0,
      avgWin,
      avgLoss,
      hasValidKelly: false,
    };
  }

  const payoffRatio = avgWin / avgLoss;
  const lossRate = 1 - winRate;
  const kellyFraction = (payoffRatio * winRate - lossRate) / payoffRatio;
  const kellyPercent = kellyFraction * 100;

  return {
    fraction: kellyFraction,
    percent: kellyPercent,
    winRate,
    payoffRatio,
    avgWin,
    avgLoss,
    hasValidKelly: true,
  };
}

/**
 * Group trades by strategy and calculate Kelly metrics for each
 */
export function calculateStrategyKellyMetrics(
  trades: Trade[]
): Map<string, KellyMetrics> {
  const strategyMap = new Map<string, Trade[]>();

  // Group trades by strategy
  for (const trade of trades) {
    const strategy = trade.strategy || "Uncategorized";
    if (!strategyMap.has(strategy)) {
      strategyMap.set(strategy, []);
    }
    strategyMap.get(strategy)!.push(trade);
  }

  // Calculate Kelly metrics for each strategy
  const metricsMap = new Map<string, KellyMetrics>();
  for (const [strategy, strategyTrades] of strategyMap.entries()) {
    metricsMap.set(strategy, calculateKellyMetrics(strategyTrades));
  }

  return metricsMap;
}
